python/benchmark/benchmark/bench_approximate_nearest_neighbors.py

```diff
@@ -0,0 +1,329 @@
+                algoParams=rt_algo_params,
+                **params,
+            ).setIdCol("id")
+
+            if is_single_col:
+                gpu_estimator = gpu_estimator.setInputCol(first_col)
+            else:
+                gpu_estimator = gpu_estimator.setInputCols(input_cols)
+
+            gpu_model, fit_time = with_benchmark(
+                "gpu fit", lambda: gpu_estimator.fit(train_df)
+            )
+
+            def transform(
+                model: ApproximateNearestNeighborsModel, df: DataFrame
+            ) -> DataFrame:
+                (item_df_withid, query_df_withid, knn_df) = model.kneighbors(df)
+                knn_df = knn_df.cache()
+                knn_df.count()
+                return knn_df
+
+            knn_df, transform_time = with_benchmark(
+                "gpu transform", lambda: transform(gpu_model, query_df)
+            )
+            total_time = round(time.time() - func_start_time, 2)
+            print(f"gpu total took: {total_time} sec")
+
+        if num_cpus > 0:
+            assert num_gpus <= 0
+
+            if is_array_col:
+                vector_df = train_df.select(
+                    "id", array_to_vector(train_df[first_col]).alias(first_col)
+                )
+
+                vector_query_df = query_df.select(
+                    "id", array_to_vector(train_df[first_col]).alias(first_col)
+                )
+
+            elif not is_vector_col:
+                vector_assembler = VectorAssembler(outputCol="features").setInputCols(
+                    input_cols
+                )
+                vector_df = vector_assembler.transform(train_df).drop(*input_cols)
+                vector_query_df = vector_assembler.transform(query_df).drop(*input_cols)
+                first_col = "features"
+            else:
+                vector_df = train_df
+                vector_query_df = query_df
+
+            if not no_cache:
+
+                (vector_df, vector_query_df), prepare_time = with_benchmark(
+                    "prepare dataset", lambda: cache_df(vector_df, vector_query_df)
+                )
+
+            from pyspark.ml.feature import (
+                BucketedRandomProjectionLSH,
+                BucketedRandomProjectionLSHModel,
+            )
+
+            cpu_estimator = BucketedRandomProjectionLSH(
+                inputCol=first_col,
+                outputCol="hashes",
+                **cpu_algo_params,
+            )
+
+            cpu_model, fit_time = with_benchmark(
+                "cpu fit time", lambda: cpu_estimator.fit(vector_df)
+            )
+
+            def cpu_transform(
+                model: BucketedRandomProjectionLSHModel,
+                df: DataFrame,
+                df_query: DataFrame,
+                n_neighbors: int,
+            ) -> None:
+
+                assert (
+                    False
+                ), "Currently does not support CPU benchmarking for approximate nearest neighbors"
```

Comment:
Shouldn't this be earlier then to avoid running fit?   Also, what is the impact of this if all benchmarks are run from a script?   Looks like this could crash the whole thing with an error condition?